[Gutenberg Data Tutorial] Part 3: Building an edit form (#38581)
WordPress/gutenberg: docs/how-to-guides/data-basics/3-building-an-edit-form.md

# Building an edit form

> This part is about adding an *Edit* feature to our app. Here's a glimpse of what we're going to build:

Reviewer: Do you want to link back to previous parts in case someone comes here first by accident?

![](./media/edit-form/form-finished.png)

### Step 1: Add an _Edit_ button

We can't have an *Edit* form without an *Edit* Button, so let's start by adding one to our `PagesList` component:

```js
import { Button } from '@wordpress/components';

const PageEditButton = () => (
	<Button variant="primary">
		Edit
	</Button>
)

function PagesList( { hasResolved, pages } ) {
	if ( ! hasResolved ) {
		return <Spinner />;
	}
	if ( ! pages?.length ) {
		return <div>No results</div>;
	}

	return (
		<table className="wp-list-table widefat fixed striped table-view-list">
			<thead>
				<tr>
					<td>Title</td>
					<td style={{width: 120}}>Actions</td>
				</tr>
			</thead>
			<tbody>
				{ pages?.map( ( page ) => (
					<tr key={page.id}>

> 						<td>{ page.title.rendered }</td>

Reviewer: Do we need to `decodeEntities` here or does `.rendered` handle that?

Author: You're right! I just added that.

						<td>
							<PageEditButton pageId={ page.id } />
						</td>
					</tr>
				) ) }
			</tbody>
		</table>
	);
}
```

The only change in `PagesList` is the additional column labeled _Actions_:

![](./media/edit-form/edit-button.png)

### Step 2: Display an _Edit_ form

Our button looks nice but doesn't do anything yet. To display an edit form, we need to have one first – let's create it:

```js
import { Button, TextControl } from '@wordpress/components';
export function EditPageForm( { pageId, onCancel, onSaveFinished } ) {
	return (
		<div className="my-gutenberg-form">
			<TextControl
				value=''
				label='Page title:'
			/>
			<div className="form-buttons">
				<Button onClick={ onSaveFinished } variant="primary">
					Save
				</Button>
				<Button onClick={ onCancel } variant="tertiary">
					Cancel
				</Button>
			</div>
		</div>
	);
}
```

> Now let's make the button display the form we just created. As this tutorial is not focused on web design, we will wire the two using a component that requires the least code: [`Modal`](https://developer.wordpress.org/block-editor/reference-guides/components/modal/). Let's update `PageEditButton` accordingly:

Reviewer: ```suggestion
Now let's make the button display the form we just created. As this tutorial is not focused on web design, we will wire the two together using a component that requires the least code: [`Modal`](https://developer.wordpress.org/block-editor/reference-guides/components/modal/). Let's update `PageEditButton` accordingly:
```

	return (
		<>
			<Button
				onClick={ openModal }
				variant="primary"
			>
				Edit
			</Button>
			{ isOpen && (
				<Modal onRequestClose={ closeModal } title="Edit page">
					<EditPageForm
						pageId={pageId}
						onCancel={closeModal}
						onSaveFinished={closeModal}
					/>
				</Modal>
			) }
		</>
	)
}
```

When you click the *Edit* button now, you should see the following modal:

![](./media/edit-form/form-scaffold.png)

Great! We now have a basic user interface to work with.

### Step 3: Populate the form with page details

We want the `EditPageForm` to display the title of the currently edited page. You may have noticed that it doesn't receive a `page` prop, only `pageId`. That's okay. Gutenberg Data allows us to easily access entity records from any component.

In this case, we need to use the [`getEntityRecord`](/docs/reference-guides/data/data-core/#getentityrecord) selector. The list of records is already available thanks to the `getEntityRecords` call in `MyFirstApp`, so there won't even be any additional HTTP requests involved – we'll get the cached record right away.

Here's how you can try it in your browser's dev tools:

```js

> wp.data.select( 'core' ).getEntityRecord( 'postType', 'page', 9 );  // Replace 9 with an actual page ID

Reviewer: Love these kinds of examples 👏

```

Let's update `EditPageForm` accordingly:

```js
export function EditPageForm( { pageId, onCancel, onSaveFinished } ) {
	const page = useSelect(
		select => select( coreDataStore ).getEntityRecord( 'postType', 'page', pageId ),
		[pageId]
	);
	return (
		<div className="my-gutenberg-form">
			<TextControl
				label='Page title:'
				value={ page.title }
			/>
			{ /* ... */ }
		</div>
	);
}
```

Now it should look like that:

![](./media/edit-form/form-populated.png)

### Step 4: Making the Page title field editable

There's one problem with our _Page title_ field: you can't edit it. It receives a fixed `value` but doesn't update it when typing. We need an `onChange` handler.

> You may have seen a pattern similar to this one in other React apps:

Reviewer: ```suggestion
You may have seen a pattern similar to this one in other React apps. It's known as a ["controlled component"](https://reactjs.org/docs/forms.html#controlled-components):
```

```js
export function VanillaReactForm({ initialTitle }) {
	const [title, setTitle] = useState( initialTitle );
	return (
		<TextControl
			value={ title }
			onChange={ setTitle }
		/>
	);
}
```

> Updating entity records in Gutenberg Data is similar. We don't use `useState`, though. The counterpart of `setTitle` is an `editEntityRecord` action which stores the updates in the Redux state. Here's how you can try it out in your browser's dev tools:

Reviewer: ```suggestion
Updating entity records in Gutenberg Data is similar but instead of using `setTitle` to store in local state, we use the `editEntityRecord` action which stores the updates in the _Redux_ state. Here's how you can try it out in your browser's dev tools:
```

I think we can simplify the wording and shorten the sentance.

```js
// Get a real page ID first

> const pageId = wp.data.select( 'core' ).getEntityRecords( 'postType', 'page' )[0].id;

Reviewer: How come you used `getEntityRecords` here and not `getEntityRecord`? Does the later not expose the ID?

Author: `getEntityRecord` does expose the ID, but it also needs one as an argument. What do I pass there? I used `getEntityRecords` to find the first available ID – it seems like this snippet could use a better explanation.

Author: Does it read better now @getdave ?

> // Uupdate the title

Reviewer: ```suggestion
// Update the title
```

wp.data.dispatch( 'core' ).editEntityRecord( 'postType', 'page', pageId, { title: 'updated title' } );
```

> At this point, you may ask _How is `editEntityRecord` this better than `useState`?_ It offers a few features you wouldn't otherwise get. We can save the changes as easily as we retrieve the data and ensure that all caches will be correctly updated. In addition, the changes applied via `editEntityRecord` are easily undo-able via the `undo` and `redo` actions. Finally,  Because the changes live in the Redux state, other components can access them. For example, we could make the `PagesList` display the currently edited title.

Reviewer: ```suggestion
At this point, you may ask _how is `editEntityRecord` better than `useState`? The answer is that it offers a few features you wouldn't otherwise get. 

Firstly, we can save the changes as easily as we retrieve the data and ensure that all caches will be correctly updated. 

Secondly, the changes applied via `editEntityRecord` are easily undo-able via the `undo` and `redo` actions. 

Lastly, because the changes live in the _Redux_ state, they are "global" and can be accessed by other components. For example, we could make the `PagesList` display the currently edited title.
```

Tweaked this. If you reject the formatting changes (no problem!) please note the grammar fixes towards the beginning.

To that last point, let's see what happens when we use `getEntityRecord` to access the entity record we just updated:

```js
wp.data.select( 'core' ).getEntityRecord( 'postType', 'page', pageId ).title
```

It doesn't reflect the edits. What's going on?

> Let's imagine it would reflect the edits. Then, anything you typed in the title field would be immediately displayed inside the `PagesList`. Gutenberg Data solves this problem by distinguishing between *Entity Records* and *Edited Entity Records*. Calling `getEntityRecord` always returns the data retrieved from the API. To see the edits applied, we need to use `getEditedEntityRecord` instead:

Reviewer: ```suggestion
Well, let's imagine for a second that it would reflect the edits. If it did, then anything you typed in the title field would be _immediately_ displayed inside the `PagesList`. 

Gutenberg Data solves this problem by distinguishing between *Entity Records* and *Edited Entity Records*. 

Calling `getEntityRecord` always returns the data retrieved _from the API_. To see the edits applied, we need to use `getEditedEntityRecord` instead:
```

Reviewer: Also I'm still not clear here as to the difference between the two. I was confused by the phrase "To see the edits applied".

Can we try to improve on the distinction between the two types of data? Is one store only in Redux state and then sync up to the API at a certain point? Is the other just whatever is on the API and ignores the Redux state?

Author: I made a clearer distinction between the two – does it make more sense now?

Reviewer: That's better thanks.

```js
// Replace 9 with an actual page ID
wp.data.select( 'core' ).getEditedEntityRecord( 'postType', 'page', 9 ).title
// 'updated title'
```

Let's update `EditPageForm` accordingly. We can access the actions using the [`useDispatch`](/packages/data/README.md#usedispatch) hook similarly to how we use `useSelect` to access selectors:

```js
import { useDispatch } from '@wordpress/data';

export function EditPageForm( { pageId, onCancel, onSaveFinished } ) {
	const page = useSelect(

> 		select => select( coreDataStore ).getEditedEntityRecord( 'postType', 'page', pageId ),

Reviewer: Right so this value comes from the Redux state so it's always "up to date" when you're editing.

If you'd run `getEntityRecord` it would show you what's currently on the server.

Question: when are the two synced up or is that a race condition kind of thing?

Author: I added the following line:

Entity records are updated to reflect any saved changes right after the REST API request is finished.

Is it clear, or do you think we should talk a bit more about that?

Reviewer: Sounds good 👍

```js
// Replace 9 with an actual page ID
wp.data.select( 'core' ).getEntityRecord( 'postType', 'page', 9 ).title
// "updated title"
```

This is how the `EditPageForm` looks like with a working *Save* button:

```js
export function EditPageForm( { pageId, onCancel, onSaveFinished } ) {
	// ...
	const { saveEditedEntityRecord } = useDispatch( coreDataStore );
	const handleSave = () => saveEditedEntityRecord( 'postType', 'page', pageId );

	return (
		<div className="my-gutenberg-form">
			{/* ... */}
			<div className="form-buttons">
				<Button onClick={ handleSave } variant="primary">
					Save
				</Button>
				{/* ... */}
			</div>
		</div>
	);
}
```

It works, but there's still one thing to fix: the form modal doesn't automatically close because we never call `onSaveFinished`. Lucky for us, `saveEditedEntityRecord` returns a promise that resolves once the save operation is finished. Let's take advantage of it in `EditPageForm`:

```js
export function EditPageForm( { pageId, onCancel, onSaveFinished } ) {
	// ...
	const handleSave = async () => {
		await saveEditedEntityRecord( 'postType', 'page', pageId );
		onSaveFinished();

> 	};

Reviewer: Nit: This is an interesting pattern. I've recently seen a problem with this whereby if the Modal is manually closed whilst the request is still running then you see an error.

Author: Good point! I'd cover that in the upcoming more advanced tutorial and for now just assume noone will close it :-) I don't want to muddy the picture too much.

Here's how we can use it in `EditPageForm`:

```js
export function EditPageForm( { pageId, onSaveFinished } ) {
	// ...
    const { lastError, page } = useSelect(
        select => ({
			page: select( coreDataStore ).getEditedEntityRecord( 'postType', 'page', pageId ),
			lastError: select( coreDataStore ).getLastEntitySaveError( 'postType', 'page', pageId )
		}),
        [ pageId ]
	)
	// ...
	return (
		<>
			{/* ... */}
			{ lastError ? (
				<div className="form-error">
					Error: { lastError.message }
				</div>
			) : false }
			{/* ... */}
		</>
	);
}
```

Great! `EditPageForm` is now fully aware of errors.

Let's see that error message in action. We'll trigger an invalid update and let it fail. The post title is hard to break, so let's set a `date` property to `-1` instead – that's a guaranteed validation error:

```js
export function EditPageForm( { pageId, onCancel, onSaveFinished } ) {
	// ...
	const handleChange = ( title ) => editEntityRecord( 'postType', 'page', pageId, { title, date: -1 } );
	// ...
}
```

Once you refresh the page, open the form, change the title, and hit save, you should see the following error message:

![](./media/edit-form/form-error.png)

> Fantastic! We can now restore the previous version of `handleChange` and move on to the next step.

Reviewer: ```suggestion
Fantastic! We can now **restore the previous version of `handleChange`** and move on to the next step.
```

# Building an edit form

This part is about adding an *Edit* feature to our app. Here's a glimpse of what we're going to build:

![](./media/edit-form/form-finished.png)

### Step 1: Add an _Edit_ button

> We can't have an *Edit* form without an *Edit* Button, so let's start by adding one to our `PagesList` component:

Reviewer: ```suggestion
We can't have an *Edit* form without an *Edit* button, so let's start by adding one to our `PagesList` component:
```

Nit: lowercase unless we're refering to the component in which case I'd recommend `<Button>`.

	}

	return (
		<table className="wp-list-table widefat fixed striped table-view-list">
			<thead>
				<tr>
					<td>Title</td>
					<td style={{width: 120}}>Actions</td>
				</tr>
			</thead>
			<tbody>
				{ pages?.map( ( page ) => (
					<tr key={page.id}>
						<td>{ decodeEntities( page.title.rendered ) }</td>
						<td>
							<PageEditButton pageId={ page.id } />
						</td>
					</tr>
				) ) }
			</tbody>
		</table>
	);
}
```

The only change in `PagesList` is the additional column labeled _Actions_:

![](./media/edit-form/edit-button.png)

### Step 2: Display an _Edit_ form

Our button looks nice but doesn't do anything yet. To display an edit form, we need to have one first – let's create it:

```js
import { Button, TextControl } from '@wordpress/components';
export function EditPageForm( { pageId, onCancel, onSaveFinished } ) {
	return (
		<div className="my-gutenberg-form">
			<TextControl
				value=''
				label='Page title:'
			/>
			<div className="form-buttons">
				<Button onClick={ onSaveFinished } variant="primary">
					Save
				</Button>
				<Button onClick={ onCancel } variant="tertiary">
					Cancel
				</Button>
			</div>
		</div>
	);
}
```

> Now let's make the button display the form we just created. As this tutorial is not focused on web design, we will wire the two together using a component that requires the least code: [`Modal`](https://developer.wordpress.org/block-editor/reference-guides/components/modal/). Let's update `PageEditButton` accordingly:

Reviewer: ```suggestion
Now let's make the button display the form we just created. As this tutorial is not focused on web design, we will wire the two together using a component that requires the least amount of code: [`Modal`](https://developer.wordpress.org/block-editor/reference-guides/components/modal/). Let's update `PageEditButton` accordingly:
```

```js
wp.data.select( 'core' ).getEntityRecord( 'postType', 'page', 9 );  // Replace 9 with an actual page ID
```

Let's update `EditPageForm` accordingly:

```js
export function EditPageForm( { pageId, onCancel, onSaveFinished } ) {
	const page = useSelect(
		select => select( coreDataStore ).getEntityRecord( 'postType', 'page', pageId ),
		[pageId]
	);
	return (
		<div className="my-gutenberg-form">
			<TextControl
				label='Page title:'
				value={ page.title }
			/>
			{ /* ... */ }
		</div>
	);
}
```

Now it should look like that:

![](./media/edit-form/form-populated.png)

### Step 4: Making the Page title field editable

There's one problem with our _Page title_ field: you can't edit it. It receives a fixed `value` but doesn't update it when typing. We need an `onChange` handler.

You may have seen a pattern similar to this one in other React apps. It's known as a ["controlled component"](https://reactjs.org/docs/forms.html#controlled-components):

```js
export function VanillaReactForm({ initialTitle }) {
	const [title, setTitle] = useState( initialTitle );
	return (
		<TextControl
			value={ title }
			onChange={ setTitle }
		/>
	);
}
```

> Updating entity records in Gutenberg Data is similar but instead of using `setTitle` to store in local state, we use the `editEntityRecord` action which stores the updates in the _Redux_ state. Here's how you can try it out in your browser's dev tools:

Reviewer: Just thinking...will people know what Redux is and how it's related to WP data?

Author: Good question! I don't intend to explain that here, but a link or two could help 👍

Author: I'm not sure how I feel about linking each mention of Redux, so I went for adding the following line to step 1:

> Also, this tutorial will lean heavily on Redux concepts such as state, actions, and selectors. If you are not familiar with them, you may want to start by reviewing [Getting Started With Redux](https://redux.js.org/introduction/getting-started).

Do you think that is sufficient?

> Updating entity records in Gutenberg Data is similar but instead of using `setTitle` to store in local state, we use the `editEntityRecord` action which stores the updates in the _Redux_ state. Here's how you can try it out in your browser's dev tools:

Reviewer: ```suggestion
Updating entity records in Gutenberg Data is similar but instead of using `setTitle` to store in local (component level) state, we use the `editEntityRecord` action which stores the updates in the _Redux_ state. Here's how you can try it out in your browser's dev tools:
```

```js
// We need a valid page ID to call editEntityRecord, so let's get the first available one using getEntityRecords.
const pageId = wp.data.select( 'core' ).getEntityRecords( 'postType', 'page' )[0].id;

// Update the title
wp.data.dispatch( 'core' ).editEntityRecord( 'postType', 'page', pageId, { title: 'updated title' } );
```

At this point, you may ask _how is `editEntityRecord` better than `useState`? The answer is that it offers a few features you wouldn't otherwise get.

Firstly, we can save the changes as easily as we retrieve the data and ensure that all caches will be correctly updated.

Secondly, the changes applied via `editEntityRecord` are easily undo-able via the `undo` and `redo` actions.

Lastly, because the changes live in the _Redux_ state, they are "global" and can be accessed by other components. For example, we could make the `PagesList` display the currently edited title.

To that last point, let's see what happens when we use `getEntityRecord` to access the entity record we just updated:

```js
wp.data.select( 'core' ).getEntityRecord( 'postType', 'page', pageId ).title
```

It doesn't reflect the edits. What's going on?

> Well, let's imagine for a second that it would reflect the edits. If it did, then anything you typed in the title field would be _immediately_ displayed inside the `PagesList`.

Reviewer: I think we need to add some wording here to clarify that the changes you type into the form field would be reflected in the PageList before you were ready to "commit" or persist them.

Author: I made some tweaks – is it clearer now?

Gutenberg Data distinguishes between *Entity Records* and *Edited Entity Records*. *Entity Records* reflect the data from the API and ignore any local edits, while *Edited Entity Records* also have all the local edits applied on top. Both co-exist in the Redux state at the same time.

Let's see what happens if we call `getEditedEntityRecord`:

```js
wp.data.select( 'core' ).getEditedEntityRecord( 'postType', 'page', pageId ).title
// "updated title"

wp.data.select( 'core' ).getEntityRecord( 'postType', 'page', pageId ).title
// { "rendered": "<original, unchanged title>", "raw": "..." }
```

> As you can see, the `title` of an Entity Record is an object, but the `title` of an Edited Entity record is a string.

Reviewer: Why do we have inconsistent response shapes? Seems like a recipe for bugs...

Reviewer: Obviously not something we can/will fix in this PR.

> This is no accident. Fields like `title`, `excerpt`, and `content` can only be rendered on the server so the REST API exposes both the `raw` markup and the `rendered` string. For example, in the block editor, `content.rendered` could used as a visual preview, and `content.raw` could be used to populate the code editor.

Reviewer: > can only be rendered on the server

How so? I wasn't clear here.

Author: Thanks for noting! I'll revise this using a few lines from the [TypeScript types PR](https://github.com/WordPress/gutenberg/pull/38666/) where this is explained more clearly.

> This is no accident. Fields like `title`, `excerpt`, and `content` can only be rendered on the server so the REST API exposes both the `raw` markup and the `rendered` string. For example, in the block editor, `content.rendered` could used as a visual preview, and `content.raw` could be used to populate the code editor.

Reviewer: ```suggestion
This is no accident. Fields like `title`, `excerpt`, and `content` can only be rendered on the server and so the REST API exposes both the `raw` markup _and_ the `rendered` string. For example, in the block editor, `content.rendered` could used as a visual preview, and `content.raw` could be used to populate the code editor.
```
